plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeProjectionPushdownPlans.java

```diff
@@ -0,0 +1,246 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.deltalake;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.trino.Session;
+import io.trino.metadata.QualifiedObjectName;
+import io.trino.metadata.TableHandle;
+import io.trino.plugin.deltalake.metastore.TestingDeltaLakeMetastoreModule;
+import io.trino.plugin.hive.metastore.Database;
+import io.trino.plugin.hive.metastore.HiveMetastore;
+import io.trino.spi.connector.ColumnHandle;
+import io.trino.spi.predicate.Domain;
+import io.trino.spi.predicate.TupleDomain;
+import io.trino.spi.security.PrincipalType;
+import io.trino.sql.planner.assertions.BasePushdownPlanTest;
+import io.trino.testing.LocalQueryRunner;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.Predicates.equalTo;
+import static com.google.common.io.MoreFiles.deleteRecursively;
+import static com.google.common.io.RecursiveDeleteOption.ALLOW_INSECURE;
+import static com.google.inject.util.Modules.EMPTY_MODULE;
+import static io.trino.plugin.hive.metastore.file.FileHiveMetastore.createTestingFileHiveMetastore;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.any;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.anyTree;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.expression;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.filter;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.join;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.project;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.tableScan;
+import static io.trino.sql.planner.plan.JoinNode.Type.INNER;
+import static io.trino.testing.TestingNames.randomNameSuffix;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static java.lang.String.format;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestDeltaLakeProjectionPushdownPlans
+        extends BasePushdownPlanTest
+{
+    private static final String CATALOG = "delta";
+    private static final String SCHEMA = "test_schema";
+
+    private File baseDir;
```

Comment:
I found that for arrays the projection pushdown is not effective:

```
create table delta.tiny.test2 (c1 integer, c2 array[row(leaf1 integer, leaf2 integer)]); 
insert into delta.tiny.test2 values (1, array(row(10,11), row(20,22));

trino> explain analyze select c2[2].leaf2 from delta.tiny.test2;
                                                                            Query Plan                                        >
------------------------------------------------------------------------------------------------------------------------------>
 Trino version: dev                                                                                                           >
 Queued: 369.02us, Analysis: 34.05ms, Planning: 46.82ms, Execution: 183.76ms                                                  >
 Fragment 1 [SOURCE]                                                                                                          >
     CPU: 4.86ms, Scheduled: 10.43ms, Blocked 0.00ns (Input: 0.00ns, Output: 0.00ns), Input: 1 row (35B); per task: avg.: 1.00>
     Output layout: [expr]                                                                                                    >
     Output partitioning: SINGLE []                                                                                           >
     ScanProject[table = delta:tiny.test2]                                                                                    >
         Layout: [expr:integer]                                                                                               >
         Estimates: {rows: 1 (5B), cpu: 55, memory: 0B, network: 0B}/{rows: 1 (5B), cpu: 5, memory: 0B, network: 0B}          >
         CPU: 5.00ms (100.00%), Scheduled: 10.00ms (100.00%), Blocked: 0.00ns (?%), Output: 1 row (5B)                        >
         Input avg.: 1.00 rows, Input std.dev.: 0.00%                                                                         >
         expr := "c2"[BIGINT '2'][2]                                                                                          >
         c2 := c2:array(row(leaf1 integer, leaf2 integer)):REGULAR                                                            >
         Input: 1 row (35B), Filtered: 0.00%, Physical input: 501B, Physical input time: 3710000.00ns                         >
                                                                                                                              >
                                                                                                                              >
(1 row)


```


This may be linked with 

```
 Map<ConnectorExpression, ProjectedColumnRepresentation> columnProjections = projectedExpressions.stream()
                .collect(toImmutableMap(Function.identity(), HiveApplyProjectionUtil::createProjectedColumnRepresentation));
```

in `DeltaLakeMetadata.applyProjection`.


The `messageColumn` in `io.trino.plugin.hive.parquet.ParquetPageSourceFactory#createPageSource(io.trino.filesystem.TrinoInputFile, long, long, java.util.List<io.trino.plugin.hive.HiveColumnHandle>, io.trino.spi.predicate.TupleDomain<io.trino.plugin.hive.HiveColumnHandle>, boolean, org.joda.time.DateTimeZone, io.trino.plugin.hive.FileFormatDataSourceStats, io.trino.parquet.ParquetReaderOptions, java.util.Optional<io.trino.parquet.ParquetWriteValidation>, int)`


```
Optional[message trino_schema {
  optional group c2 (LIST) {
    repeated group list {
      optional group element {
        optional int32 leaf1;
        optional int32 leaf2;
      }
    }
  }
}
]
```

(contains `leaf1` even IMO though it shouldn't).

Reply:
If this is valid - pls weight whether this should be addressed in this PR (see if it happens for hive as well) or it should be addressed as a follow-up.

Reply:
Dereference pushdown for array and map type into the connector is not supported currently, it's a separate topic for improvement. See `ConnectorExpressionTranslator#visitSubscriptExpression`